### PR TITLE
Basic sysl logger library

### PR DESCRIPTION
**Pull request by nofun97**
Part of #397 

Changes proposed in this pull request:
- Added a basic library for logging that enforces the use of context
- Added an underlying logger (logrus Standard Logger)

Further work required:
1. Adding test logger (using logrus null logger and hook feature to replace existing log tests within sysl)
2. Adding codelinker to the library so that each logging message can link itself to the part of the code where the log was executed
3. Adding `Configure` feature
4. Replacing logrus in sysl
5. Add colors to the messages (not a priority unless requested)

@anz-bank/sysl-developers

### CODE REVIEW

File: pkg/syslLogger/api_test.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+			}
+		}(c))
+	}
+}
+
+func testAllCases(t *testing.T, name string, test func(testutil.LogCases, *testing.T) func(*testing.T)) {
+	for _, c := range testutil.GenerateLogCases() {
+		t.Run(name+" "+c.Name, test(c, t))
+	}
+}
```
Comment (anzdaddy) [inline pkg/syslLogger/api_test.go]: Simpler:
```go
func testAllCases(t *testing.T, name string, test func(testutil.LogCases, *testing.T)) {
	for _, c := range testutil.GenerateLogCases() {
		t.Run(name+" "+c.Name, func(tt *testing.T) { test(c, tt) })
	}
}
```

File: pkg/syslLogger/testutil/testutil.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
+package testutil
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/testutil/testutil.go]: If I put it in the `loggers` package, it will expose the `testutil` functions to users. Since the main API and the `loggers` package use it, I thought it would be neater to put it in a separate package.

File: pkg/syslLogger/loggers/mockLogger.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	arguments := append([]interface{}{format}, args...)
+	m.Called(arguments...)
+}
+
+func (m *MockLogger) Panic(args ...interface{}) {
+	m.Called(args...)
+}
+
+func (m *MockLogger) Panicf(format string, args ...interface{}) {
+	arguments := append([]interface{}{format}, args...)
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/loggers/mockLogger.go]: using `m.Called(format, args...)` gave me a syntax error

File: pkg/syslLogger/testutil/testutil.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (anzdaddy) [reply pkg/syslLogger/testutil/testutil.go]: That's cool. I wasn't sure whether it was used by one package or more.

File: pkg/syslLogger/loggers/loggerInterface.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	Panic(args ...interface{})
+	Panicf(format string, args ...interface{})
+	Print(args ...interface{})
+	Printf(format string, args ...interface{})
+	Trace(args ...interface{})
+	Tracef(format string, args ...interface{})
+	Warn(args ...interface{})
+	Warnf(format string, args ...interface{})
+	WithField(key string, val interface{}) Logger
+	WithFields(fields map[string]interface{})
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/loggers/loggerInterface.go]: The idea for `WithField` to return a `Logger` is so that user can do a chain to add the fields. Something like this.
```go
logger.WithField("val1", 1).WithField("val2", 2).WithField("val3", 3).Debug("This is a message")
```

So for `WithFields` to return a `Logger` seems pointless. Would it be better to return the `Logger` anyway?

[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/loggers/loggerInterface.go]: Will change it to return a new `Logger`

File: pkg/syslLogger/api.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+import (
+	"context"
+
+	"github.com/anz-bank/sysl/pkg/syslLogger/loggers"
+)
+
+type loggerContextKey int
+
+const loggerKey = loggerContextKey(1)
```
[the remark this answers is not in the document]
Comment (anz-gordonj7) [reply pkg/syslLogger/api.go]: I think I read that for context keys just doing `type loggerKey struct {}`is more idiomatic?

File: pkg/syslLogger/loggers/loggerInterface.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+	Fatalf(format string, args ...interface{})
+	Panic(args ...interface{})
+	Panicf(format string, args ...interface{})
+	Print(args ...interface{})
+	Printf(format string, args ...interface{})
+	Trace(args ...interface{})
+	Tracef(format string, args ...interface{})
+	Warn(args ...interface{})
+	Warnf(format string, args ...interface{})
+	WithField(key string, val interface{}) Logger
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/loggers/loggerInterface.go]: No, will change it to `PutField` and `PutFields` to avoid confusion

File: pkg/syslLogger/api_test.go (hunk 1 above)
Comment (nofun97) [reply pkg/syslLogger/api_test.go]: it was done that way to avoid `scopelint` issues

File: pkg/syslLogger/loggers/loggerInterface.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (anz-gordonj7) [reply pkg/syslLogger/loggers/loggerInterface.go]: No, leave it as `With...` and change the implementation to not affect the current logger please :)

[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/loggers/loggerInterface.go]: The thing is, `WithField` returns a logger so that it can chain fields insertion like this
```go
logger.WithField("val1", 1).WithField("val2", 2).WithField("val3", 3).Debug("This is a message")
```

If `WithField` copies the logger every time, it would be really slow. My plan was to just invoke `Copy` in the main API.

Is that better or returning a new `Logger` would be better?

[the remark this answers is not in the document]
Comment (anz-gordonj7) [reply pkg/syslLogger/loggers/loggerInterface.go]: 1. "it would be really slow" is a banned phrase.... PROVE IT!!
2. That is nonsensical code anyway, the caller should be doing `logger.WithFields(map[string]intereface{}{"val":1, "val2:2....`

File: pkg/syslLogger/api.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+func AddLogger(ctx context.Context, logger loggers.Logger) context.Context {
```
[the remark this answers is not in the document]
Comment (anz-gordonj7) [reply pkg/syslLogger/api.go]: or AsContext()?

File: pkg/syslLogger/api.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+	return context.WithValue(ctx, loggerKey, logger.Copy())
+}
+
+/**
+ * The fields setup in AddField and AddFields are for context specific fields
+ * Fields will be logged alphabetically
+ */
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/api.go]: fixed

File: pkg/syslLogger/api_test.go [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+func TestAddField(t *testing.T) {
+	cases := testutil.GenerateSingleFieldCases()
+	for _, c := range cases {
+		t.Run(c.Name, func(sc testutil.SingleField) func(*testing.T) {
+			return func(tt *testing.T) {
+				tt.Parallel()
+
+				logger := loggers.NewMockLogger()
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/api_test.go]: I added something similar to it as there's a bit of main API changes

File: pkg/syslLogger/api_test.go [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+	ctx := context.Background()
+	logger := loggers.NewMockLogger()
+	logger.On("Copy").Return(loggers.NewMockLogger()).Once()
+	newCtx := AddLogger(ctx, logger)
+	newLogger := newCtx.Value(loggerKey).(*loggers.MockLogger)
+
+	require.True(t, logger.AssertExpectations(t))
+	require.True(t, &ctx != &newCtx)
+	require.True(t, &logger != &newLogger)
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/api_test.go]: replaced some with `assert`

File: pkg/syslLogger/api.go [hunk 11]
```diff
@@ -0,0 +1,101 @@
+package sysllogger
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/api.go]: changed to `log`

File: pkg/syslLogger/api.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/api.go]: using `iota` for now

File: pkg/syslLogger/api.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+func AddLogger(ctx context.Context, logger loggers.Logger) context.Context {
+	return context.WithValue(ctx, loggerKey, logger.Copy())
+}
+
+/**
+ * The fields setup in AddField and AddFields are for context specific fields
+ * Fields will be logged alphabetically
+ */
+func AddField(ctx context.Context, key string, val interface{}) context.Context {
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/api.go]: fixed

File: pkg/syslLogger/api.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+	newLogger := getLogger(ctx)
+	newLogger.WithField(key, val)
+	return AddLogger(ctx, newLogger)
+}
+
+func AddFields(ctx context.Context, fields map[string]interface{}) context.Context {
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/api.go]: fixed

File: pkg/syslLogger/loggers/loggerInterface.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/loggers/loggerInterface.go]: changed to `PutField` and `PutFields`

File: pkg/syslLogger/loggers/loggerInterface.go [hunk 14]
```diff
@@ -0,0 +1,22 @@
+package loggers
+
+type Logger interface {
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/loggers/loggerInterface.go]: added docs

File: pkg/syslLogger/api.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+	newLogger := getLogger(ctx)
+	newLogger.WithFields(fields)
+	return AddLogger(ctx, newLogger)
+}
+
+func LogField(ctx context.Context, key string, value interface{}) loggers.Logger {
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/syslLogger/api.go]: added docs

File: pkg/syslLogger/api_test.go (hunk 1 above)
Comment (nofun97) [reply pkg/syslLogger/api_test.go]: fixed

File: pkg/log/loggers/standardLogger.go [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+	case logrus.DebugLevel:
+		return "DEBUG"
+	case logrus.TraceLevel:
+		return "TRACE"
+	case logrus.WarnLevel:
+		return "WARN"
+	case logrus.ErrorLevel:
+		return "ERROR"
+	case logrus.InfoLevel:
+		// Print logs at the InfoLevel
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/log/loggers/standardLogger.go]: It's how logrus process it. If you do `Logrus.Info("message")` and `Logrus.Print("Message")` both will log at the same level which is `logrus.InfoLevel`, just the message is logged slightly different (`Info` will log `[message]` and `Print` will log `message`). For simplicity sake, I just removed `Info` and replace it with `Print`.

[the remark this answers is not in the document]
Comment (anzdaddy) [reply pkg/log/loggers/standardLogger.go]: Wouldn't it be simpler to remove `Print` instead?

File: pkg/log/api.go [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+// The fields setup in WithField and WithFields are for context specific fields
+// Fields will be logged alphabetically
+
+// WithField adds a single field in the scope of the context
+func WithField(ctx context.Context, key string, val interface{}) context.Context {
+	newLogger := getLogger(ctx).Copy()
+	newLogger.PutField(key, val)
+	return context.WithValue(ctx, loggerKey, newLogger)
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/log/api.go]: fixed

File: pkg/log/loggers/loggerInterface.go [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+	Trace(args ...interface{})
+	// Tracef logs the message at the Trace level
+	Tracef(format string, args ...interface{})
+	// Warn logs the message at the Warn level
+	Warn(args ...interface{})
+	// Warnf logs the message at the Warn level
+	Warnf(format string, args ...interface{})
+	// PutField returns the Logger with the new field added
+	PutField(key string, val interface{}) Logger
+	// PutFields returns the Logger with the new fields added
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/log/loggers/loggerInterface.go]: now `PutFields` return the `Logger`

File: pkg/log/loggers/standardLogger.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/log/loggers/standardLogger.go]: `Logrus.Info("this is a message")` logs would appear like
```
(timestamp) info [this is a message]
```
while `Logrus.Print` would do the same thing without the `[]`. Would it be better to do `return "PRINT"` instead?

[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/log/loggers/standardLogger.go]: fixed, replaced `parseLevel` with directly converting logrus level string to uppercase
```go
strings.ToUpper(entry.Level.String())
```

File: pkg/log/api_test.go [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+func TestWithField(t *testing.T) {
+	testSingleFieldCases(t, "TestWithField",
+		func(tt *testing.T, ctx context.Context, sc testutil.SingleField) {
+			logger := ctx.Value(loggerKey).(*loggers.MockLogger)
+			newCtx := WithField(ctx, sc.Key, sc.Val)
+			newLogger := newCtx.Value(loggerKey).(*loggers.MockLogger)
+
+			require.True(tt, logger.AssertExpectations(tt))
+			assert.True(tt, &ctx != &newCtx)
+			assert.True(tt, &logger != &newLogger)
```
Comment (anzdaddy) [inline pkg/log/api_test.go]: You are comparing the addresses of variables, which will never be equal.

File: pkg/log/api.go [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+
+// Field adds a single field for log specific fields. Does not add to the context.
+func Field(ctx context.Context, key string, value interface{}) loggers.Logger {
+	return getLogger(ctx).Copy().PutField(key, value)
+}
+
+// Fields adds multiple fields for log specific fields. Does not add to the context.
+func Fields(ctx context.Context, fields map[string]interface{}) loggers.Logger {
+	return getLogger(ctx).Copy().PutFields(fields)
+}
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/log/api.go]: it was a feature suggested by @anzboi

File: pkg/log/api_test.go (hunk 19 above)
Comment (nofun97) [reply pkg/log/api_test.go]: after looking through the tests again, I realized that those are values of returned from a mock interface. It might be pointless to test that, will remove these tests.

File: pkg/log/api.go [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+type loggerContextKey int
+
+const loggerKey loggerContextKey = iota
+
+// WithLogger adds a copy of the logger to the context
+func WithLogger(ctx context.Context, logger loggers.Logger) context.Context {
+	return context.WithValue(ctx, loggerKey, logger.Copy())
+}
+
+// The fields setup in WithField and WithFields are for context specific fields
```
[the remark this answers is not in the document]
Comment (nofun97) [reply pkg/log/api.go]: fixed

File: pkg/log/loggers/loggerInterface.go [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+	// Trace logs the message at the Trace level
+	Trace(args ...interface{})
+	// Tracef logs the message at the Trace level
+	Tracef(format string, args ...interface{})
+	// Warn logs the message at the Warn level
+	Warn(args ...interface{})
+	// Warnf logs the message at the Warn level
+	Warnf(format string, args ...interface{})
+	// PutField returns the Logger with the new field added
+	PutField(key string, val interface{}) Logger
```
Comment (anzboi) [inline pkg/log/loggers/loggerInterface.go]: Normally a chaining method is called `With`, is this to avoid a name clash?

Comment (nofun97) [reply pkg/log/loggers/loggerInterface.go]: `PutField` modifies the original logger, if I named it `WithField` it implies that the original logger is not modified as discussed with @anzdaddy. I chose to modify the original logger so that each `PutField` in a chained operation does not copy the logger many times. The main API will enforces the copying of loggers.

**Comment by nofun97** [comment]
moved to [pkg](https://github.com/anz-bank/pkg/pull/1)
